src/plugins/data_explorer/public/utils/state_management/preload.ts

```diff
@@ -0,0 +1,42 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+import { PreloadedState } from '@reduxjs/toolkit';
+import { getPreloadedState as getPreloadedMetadataState } from './metadata_slice';
+import { RootState } from './store';
+import { DataExplorerServices } from '../../types';
+
+export const getPreloadedState = async (
+  services: DataExplorerServices
+): Promise<PreloadedState<RootState>> => {
+  let rootState: RootState = {
+    metadata: await getPreloadedMetadataState(services),
+  };
+
+  // initialize the default state for each view
+  const views = services.viewRegistry.all();
+  const promises = views.map(async (view) => {
+    if (!view.ui) {
+      return;
+    }
+
+    const { defaults } = view.ui;
+
+    // defaults can be a function or an object
+    const preloadedState = typeof defaults === 'function' ? await defaults() : defaults;
+    rootState[view.id] = preloadedState.state;
+
+    // if the view wants to override the root state, we do that here
+    if (preloadedState.root) {
+      rootState = {
+        ...rootState,
+        ...preloadedState.root,
+      };
```

Comment:
These guys are asynchronously updating `rootState`. We could end up with some overwriting each other, producing unexpected and different results. Are we good with that?

Reply:
Hmm, good call. I need this because saved searches also specify the index pattern that we need on page load and that's in the root state. I am worried about the non deterministic nature of the load though even if the overwrite is fine with me. Any suggestions on how to improve this? I do want to avoid serializing this if possible because it increases page load time. This can also be improved in future l, but I do see the problem